### PR TITLE
fix: rename and fetch_data progress report

### PR DESCRIPTION
**Pull request by ho-229**

### CODE REVIEW

File: src/filter/proxy.rs [hunk 1]
```diff
@@ -232,8 +232,12 @@ pub unsafe extern "system" fn notify_rename<T: SyncFilter + 'static>(
     if let Some(filter) = filter_from_info::<T>(info) {
         let request = Request::new(*info);
         let ticket = ticket::Rename::new(request.connection_key(), request.transfer_key());
+        let info = info::Rename(
+            (*params).Anonymous.Rename,
+            request.volume_letter().to_os_string(),
+        );
```
Comment (ok-nick) [inline src/filter/proxy.rs:238]: Ah, this is so ugly but it must be done.

File: examples/sftp/src/main.rs [hunk 2]
```diff
@@ -264,44 +220,32 @@ impl SyncFilter for Filter {
...
                 }
-                // TODO: do I need to delete it locally?
-                false => self
-                    .sftp
-                    .unlink(Path::new(unsafe {
-                        OsStr::from_encoded_bytes_unchecked(request.file_blob())
-                    }))
-                    .map_err(|_| CloudErrorKind::InvalidRequest)?,
+                (true, false) => {
+                    fs::copy(&src, dest).map_err(|_| CloudErrorKind::InvalidRequest)?;
```
[the remark this answers is not in the document]
Comment (ho-229) [reply examples/sftp/src/main.rs:240]: Use `fs::rename` here will cause `ERROR_SHARING_VIOLATION`.

File: examples/sftp/src/main.rs [hunk 3]
```diff
@@ -264,44 +220,32 @@ impl SyncFilter for Filter {
...
+                    self.sftp
+                        .unlink(&src.strip_prefix(&base).unwrap())
+                        .map_err(|_| CloudErrorKind::InvalidRequest)?;
+                }
+                (false, true) => Err(CloudErrorKind::NotSupported)?, // TODO
```
[the remark this answers is not in the document]
Comment (ho-229) [reply examples/sftp/src/main.rs:245]: No, it's simpler way.

File: src/filter/proxy.rs (hunk 1 above)
Comment (ho-229) [reply src/filter/proxy.rs:238]: Cause I do the same thing in the Placeholder::path

**Comment by ho-229** [comment]
Hi @ok-nick, any problems left?

File: examples/sftp/src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ok-nick) [reply examples/sftp/src/main.rs:245]: I agree, a lot of the impl can be greatly simplified and probably replaced with a single call to `fs::copy`, but this functionality shouldn't be removed from the example.

**Comment by ho-229** [comment]
This fix looks weird but it works well.

`Cloud Filter` will copy remote files(or directories) to local automatically. However, `renamed` will not be invoked after `rename` so we can not remove the src from the remote.

**Comment by ok-nick** [comment]
Ah, so when the target isn't in scope and the source is, it will invoke `SyncFilter::delete` rather than `SyncFilter::rename`?

If that's the case, I wonder if it'd be better to replace the `source_in_scope`/`target_in_scope` fns with a single enum to make it more clear. Something like: `RenameAction::MoveInsideFromOutside` + `RenameAction::MoveInsideFromInside`.

**Comment by ho-229** [comment]
It will not inoke `SyncFilter::delete` in this case.

**Comment by ok-nick** [comment]
How is the file removed from the remote then, if not invoked by `SyncFilter::delete`? I also don't understand how `SyncFilter::renamed` plays a role here?

Also, we should keep the existing functionality for moving a file from outside the sync filter to inside the sync filter rather than keeping it as `CloudErrorKind::NotSupported` [here](https://github.com/ok-nick/wincs/pull/12#discussion_r1640447120).

**Comment by ok-nick** [comment]
I pushed a branch named `update-deps` to update all the dependencies, let me know if it works for you: https://github.com/ok-nick/wincs/tree/update-deps

**Comment by ho-229** [comment]
I think I can't fix this problem, thanks for your reviews

**Pull request closed without merging**